kotlinx-coroutines-core/common/src/JobSupport.kt

```diff
@@ -916,16 +942,26 @@ public open class JobSupport constructor(active: Boolean) : Job, ChildJob, Paren
     private fun firstChild(state: Incomplete) =
         state as? ChildHandleNode ?: state.list?.nextChild()
 
-    // return false when there is no more incomplete children to wait
-    // ## IMPORTANT INVARIANT: Only one thread can be concurrently invoking this method.
-    private tailrec fun tryWaitForChild(state: Finishing, child: ChildHandleNode, proposedUpdate: Any?): Boolean {
+    /**
+     * This method is invoked by the JobSupport lifecycle when it transitions into "completing" state (also includes "cancelling"),
+     * and in order to be transitioned into its final state it has to wait for all the children.
+     * If this method returns `true` then the current job is waiting for its children.
+     * It's done via an addition of a special completion handler to the child that, when invoked, invokes
+     * [continueCompleting] that almost immediately fallbacks to [tryWaitForChildren].
+     *
+     * If it returns `false` then there is no children to wait for and it's safe to finalize the final job's state.
+     * Note that at this point, new children can appear:
+     * * For `cancelling` state they are immediately cancelled.
+     * * For `completing` state they are not cancelled and left hanging out to dry.
```

Comment:
I wanted to write something meaningful here, but failed to do so.

Here is a more formal description of what it means: https://github.com/Kotlin/kotlinx.coroutines/pull/3890